### ISSUE TITLE
Diversified H2H Equipment Behaviour in Boarding and Capture

### ISSUE THREAD

**Issue by MasterOfGrey** [endless-sky/endless-sky#5425]
### Is your feature request related to a problem? Please describe.
Boarding is a simple activity, and while no one is keen to turn it into some damn mini-game, the current nature is such that there are H2H options and tactics that are simply *always* superior.
The intent of this is to propose a relatively low-effort change which would address the lack of diversity and still ensure that the *work* involved in capturing is still done in preparation and planning, rather than making the actual act of capturing a more drawn out and exhausting process.

### Related Issue Links
 There is some derived content from here: [Issue 2392](https://github.com/endless-sky/endless-sky/issues/2392)

### Describe the solution you'd like
Essentially this becomes a system of attributes that apply modifiers to the base strength of different H2H weapons.
These modifiers either exists as part of the ship template, or as things added by particular outfits.

**Your first set of modifiers are Ship terrain:**
Corridor Width:
- Narrow
- Standard
- Wide
- Open
(These range from one individual passing at a time in a budget vessel like the Scout, to attacking something like a worldship that is likely to have big open spaces both for storage, mass transit and recreation purposes.)

Design Layout:
- Twisted
- Straight

Ventilation Capacity:
- Poor
- Adequate
- Good

**Your second set of modifiers are Ship/Crew Defences:**
Defences:
- Ventable Sections
- Combat Environmental Suits
- Defensive Alcoves
- Security Embrasures

**Impact on Weapons:**
These modifiers then affect the base strength of weapons. 
All modifiers are conceived of as negative for the principle case. Meaning, a H2H outfit's best effectiveness is it's base stats, circumstances that counter it provide penalties.
It is coneivable that a weapon could receive a positive modifier for something, but no good examples of this spring to mind.
However, the effect of the modifier should be multiplicative/additive in a way described below.

Examples:
- Nerve Gas will be fully effective against narrow corridors, with a twisted layout, and poor ventilation
- - It will be less effective under the inverse, and additionally reduced by the ship having the defensive qualities: 
- - - Ventable Sections
- - - Combat Environmental Suits
.
- Laser Rifles will be fully effective against standard corridors with straight layout.
- - They will have no modifier from ventilation capacity.
- - They will be less effective the further from standard they get and with twisted layout, and additionally reduced by the ship having the defensive qualities: 
- - - Defensive Alcoves
- - - Security Embrasures

Things like grenades might also have the `Consumable` tag, so that if those weapons are the ones being prioritised, then every casualty on the enemy side would subtract 1 from the consumable stash, potentially causing crew to switch to other weapons from one rounf to the next as supplies dwindle.

**Some current H2H equipment would then become a system, which you only install one of on a ship:**
i.e.
Security Station:
- Adds the attribute "Security Embrasures" to your ship.

Combat Environmental Suits: (or any other powered suit)
- Adds the attribute "Combat Environmental Suits" to your ship.
- Adds upkeep costs equal to X multiplied by the number of crew on board.

Reinforced Bulkhead Cotrols:
- Adds the attribute "Ventable Sections" to your ship.

**The maths:**
It should not be possible to add two systems that add the same modifiers onto the same ship.
However, a single system should be able to add the same modifier more than once.
So adding *more* embrasures doesn't work (you've only got limited spaces), but adding *better* embrasures does.

Hence if a weapon has a -0.1 penalty for a particular attribute, the penalty from this attribute is:
Base Strength x Penalty

If the Attribute is applied twice the penalty is:
(Base Strength x Penalty) + ((Base Strength -(Base Strength x Penalty)) x Penalty)

After which the penalties from different attributes are added together:
Penalty 1 + Penalty 2 + Penalty 3 = Total Penalty

### A single tweak to the existing mechanics: (optional but desirable)
**Rather than 2 Actions have 3:**
- Attack: As per normal, your forces attack the enemy.
- Hold Ground: Your forces defend against the enemy. They will do damage if the enemy picked attack, but no damage if the enemy picked hold ground or retreat.
- - (This is like the current defend action, but combat doesn't end if you both defend.)
- Retreat: Your forces defend against the enemy and pull back to your ship. 
- - Damage you inflict is halved.
- - The battle ends if attacker picks retreat and defender picks hold ground/defend.
- - If attacker picks retreat and defender picks attack, defender does double damage to the retreating choice.
- - The combat immediately ends after this unless the original defender has 2x the combat strength of your remaining crew, in which case the current behaviour of them continuing to attack you proceeds.

### Additional Cool Feature Option
**Combat Width:**
How wide and how twisted the corridors are for the ship that is being boarded impacts the number of crew that can fight at once.
This would simply serve to Truncate the list of crew and their equipment being factored into the current calculation for strength when casualties are resolved.

### UI Presentation
Rather than reworking, simply expanding on the existing UI could facilitate this.

**On either side of the existing box:**
A window displaying all attributes that are currently applied. E.g.:
- Narrow
- Twisted
- Good
- Ventable Sections
- Defensive Alcoves

**On a window above the resulting wider window:**
Have a little field that shows:
- how many of your crew will be engaged in the next round of combat
- how many of each H2H weapon they'll be using. (i,e, 25 crew combat width, 15 pug sticks, 10 nerve gas grenades)
- remaining supplies of H2H gear

### Conclusion
By doing this the current action of capturing a ship remains almost identical, but more diverse options are implemented to make capturing different ships a more unique experience on the planning and preparation side of things. 

### Some Ideas Posted in the Discord by Warlord Mike when the topic was initially discussed, copied without review:

ASSAULT WEAPONS: All of the following require Power Armor or Battle Suits for use, so don't bother getting them if you don't have them. The difference between Assault Weapons and Regular Weapons is that these are strictly designed for offense - they have next to zero defense power.

Heavy Machine Guns: "I am Heavy Weapons Guy...and this is my weapon." These bullet/laser/whatever hoses will quickly paint the walls with whatever the color of your enemies' blood happens to be. Reaps a grim harvest in tight corridors, but turns into peashooters in wide corridors. Effective at all ranges, otherwise.

Gauss Rifle: Sir Isaac is still the deadliest son of a bitch in space. Works best in tight, long corridors, where there's little room to maneuver and lots of targets. It loses out on potency the wider the space gets.

Tesla Cannon: The fury of the heavens at your very fingertips. Works well in any spacing, but only good up to about mid-range. Any power-armor users, however, will take double damage.

Vulcan Guns: Never enough dakka. Ever. These minigun-type weapons are even more powerful than Heavy Machine Guns, and thus are marked up accordingly by price. While the damage output is slightly weaker than Heavy Machine Guns, Vulcans will mulch enemies equally well regardless of range or spacing. 

Void Cannon: Dividing by zero never looked so cool. Creates a microscopic singularity at the point of impact, crushing its targets. Can't use it up close (unless you want to become part of the singularity, yourself), and wide corridors also hamper its use.

Fusion Cannon: Gives "the nuclear option" a whole new shine. Restricted to long-range only, this devastating weapon unleashes a small fusion explosion at the point of impact, vaporizing almost anything in its radius. Any spacing works, though tighter spacing may lead to hull integrity instabilities...

DEFENSES:

Machine Gun Nest: These portable pillboxes will turn tight corridors (and even regular ones) into bloodbaths. They lose some effectiveness in wide corridors, though.

Gas Dispensers: Stun gas is allowable; nerve gas is not. But nerve gas will also clear the rooms much faster and with more permanence...

Security Stations: You know 'em, you hate 'em. These bulkheads will withstand small-arms fire without much problem, but get ready to fall back if they bring out the big guns.

Bunkers: Significantly more armored than Security Stations, and even more effective. They take up so much room in corridors, however, it penalizes your attack, somewhat.

Auto-turrets: Sometimes the best defense is appearing to not have one at all until it's too late. These automated defenses can quickly stifle boarding actions in tight corridors. Less so in wide spaces.

Smoke Dispensers: Can't hit what isn't there to begin with. Not only does this hinder the enemies' ability to attack, but explosive and fire-based weapons get a frightening boost to their damage (think of a spark in a grain silo).

Interior Turrets: Smile for the birdie. Some crazy fools decided to take some tank turrets and use them as defensive emplacements. Armed with gatling guns and an 80mm cannon, this thing will destroy everything that isn't equipped with power armor, and even then, it'll put up a good resistance. The one downside is that your attack takes a nosedive. Off a cliff. In a dumpster. On fire.

EDITOR'S NOTE: Yes, I realize this one is absolutely batshit insane, but why not have some fun with this?

Electrified Wall Panels: Cruel and Unusual? Maybe. But it's certainly effective, especially against Powered Armor, which fries their circuitry. Beware, though: if your enemies are wearing Tesla Armor, it renders this defense useless...and boosts their attack.

ARMOR:

Leather: It's better than nothing, certainly, but won't do much against anything bigger than pistols.

Ceramic: A step in the right direction, but you can still do better. You won't need to worry about pistols and even some submachine guns, but rifles might give you pause.

Combat Armor: The baseline for crew defense. This will provide adequate defense against most small-arms fire, but won't do much against heavy artillery.

Flak Jackets: Such a simple design, and yet so very effective. Good at blocking explosives and pistol/shotgun fire; not so much against flamethrowers or sniper rifles.

Tesla Armor: While not much more effective than Combat Armor, it renders Electrified Wall Panels useless...and boosts your attack. Not a bad deal, if you ask me.

Assault Armor: Combat Armor, but better. It will provide adequate defense against even explosive rounds, and small arms are all but useless against this. But don't expect it to save you from nerve gas or electric shocks...

Powered Armor: Now we're talkin'. These armored exoskeletons will shrug off almost anything the enemy can throw at you...just stay away from Electrified Wall Panels, or you'll be flash-fried instantly.

Centurion Battle Suit: They don't get any better than this. The Centurion uses advanced nano-weave technology to provide unmatched protection, blocking anything short of a Gauss Rifle bolt. And because the suits are self-sealing and non-conductive, gas and electricity have no effect. If you died in this, you were in way over your head.

**Comment by WarlordMike** [comment]
You left out the original weapons, sir. =P

Fortunately, I have them on hand:

Melee Weapons: Sometimes, there's nothing quite as satisfying as plunging your blade into the enemy's stomach. Naturally, only useable at close range; smaller weapons (e.g., daggers) might allow more hits than larger ones (e.g., war axes).

Pistols: Useful in close-quarters, but midrange is something to be desired; near-worthless at long range.

Submachine guns: A step up from pistols - good at close range and even somewhat effective midrange, but again, useless at long range.

Carbines: Balanced at all ranges, but has less attack power than dedicated attack platforms.

Single-shot Rifles: You'll get mulched at close range, but mid-range is decent; long-range, however, is where you shine (think sniper rifles).

Assault Rifles: Burst fire weapons that will absolutely shred anything in close- or mid-range, but good luck hitting anything beyond that.

Shotguns: "Alright you primitive screwheads...listen up! _THIS_ is my _**BOOMSTICK!**_" *Evil Dead* and Ash Williams jokes, aside, these sweepers work best in tight corridors up to mid-range. They'll even work decently in regular corridors, but their performance suffers in wide corridors and at long range.

Hand Grenades: Lots of fun at parties and in wide-open spaces, but can only hit as far as your throwing arm (mid-range). Don't even bother at long range.

Rocket/Missile Launchers: These things have a close range and tight quarters "Dead Zone" (don't want to be blowing ourselves up, now do we?), but at mid- and long-range (and regular corridor sizes and larger), they'll annihilate the competition.

Flamethrowers: Yes, these things are legal to own. And in tight quarters, it'll barbecue your enemies until they're nice and crispy. Just hope they don't have long corridors, as these napalm-sprayers only hit as far as mid-range.

Grenade Launchers: Give those hand grenades some much-needed range with these bad boys. And since they're not impact-detonated, they're great at any range. Just be careful not to blow yourself up in tight quarters...

===

All this aside, looking forward to what happens!

**Comment by gunqqer** [comment]
Range doesn't really effect pistols that much, at least, the ranges that 
will be on ships. I doubt you'll see a 50 yard hallway with no 
obstructions. Also, shotguns, unlike most video games show, have a 
sizable range so unless you have a 2in barrel, you wont have a big 
enough spread to make is less effective at ship long ranges. And assault 
rifles? Burst rifle will easily get mid-range, and switch to select fire 
and long range is effective. It is a rifle.

tl;dr range isn't an issue with firearms in regards to on-ship combat.

What would matter more imo is cover and things that negate cover. Good 
luck hiding behind something when an explosive detonates beside you. Or gas.

**Comment by MasterOfGrey** [comment]
Sometimes it's not just range, a laser rifle is a long weapon, it is likely to be unwieldy to bring to bear in very close quarters, and as a very precise weapon it is comparatively less effective than something with a splash effect in a wide open area where your targets could be in many directions.

Ultimately though those are merely examples. I am not married to any particular numbers, I'm just demonstrating the system.

**Comment by gunqqer** [comment]
I think it should be split into H2H range, and everything else. Maybe a 
midrange for grenades and launcher weapons.

**Comment by beccabunny** [comment]
I admit that the first time I saw this I was against the change, but after a bit of thinking I changed my mind. Not commenting on the actual options yet, I believe that this could happen this way, requiring the least amount of changes:
- the stats are applied to the single ship stats, and as such they are visible in the shipyard view;
- the outfits have specific weaknesses in their data in numeric values, which are instead not shown in the data but are mentioned in their description (since the only time they are relevant is when you are looking at them in the outfitter)
- the only way to see a ship's defensive stats without having access to a shipyard that sells it is through an outfit scanner: I chose this because I was not sold of the idea of adding another scanner entirely; the tactical scanner is already quite useful and does not have scanning times; and the outfit scanner already does something similar to what would be required to have this informations
- in the boarding UI, the only change is a phrase saying that "The effectiveness of an attack can vary depending on the opposite ship's defensive values."

**Comment by MasterOfGrey** [comment]
@gunqqer there are as many types of alien weapons as there are imaginations to develop them

This is why I went with ship attributes, the people putting stats and modifiers to weapons can decide whether or not, and to what degree, they’re affected by those things.

**Comment by gunqqer** [comment]
Oh yeah having a varied amount of attributes is a good idea, it is the 
who "short barreled guns have a very limited range" sort of thing I want 
to avoid. I want the weapons to make sense in a realistic sort of way.

**Comment by ravenshining** [comment]
If all this nerfs my "start in a shuttle and fast-track my way to insane wealth by capping interceptors" strategy, and slows down cap-per-day rates overall by forcing players to plan a capture, head to an outfitter, reconfigure, etc., without affecting balance on a potential-target-defence-per-attackers-bunk basis, it'll get my vote.

**Comment by beccabunny** [comment]
Regarding the sintax, I was thinking about having the attribute followed by either 0 or 1 if there are only two choices (for example, `ventable sections 1` for ships with ventable sections) or a decimal value between 0 and 1 if there are more than two options (`ventilation capacity 0` to `ventilation capacity 0.3` gets translated in game as "bad ventilation capacity"). The same happens for the hand to hand outfits, only that they always have decimal values (to add another exampe to the ventilation list, `venting 0.9` would be a weapon that gets greatly nerfed by the venting capacity of a ship). Then, the formula would be something akin to `attack-=(attack*ship defensive attribute*outfit respective attribute)`, applied one time for each of these attributes. For example, using nerve gas with a venting of 0.8 and on a ship with a bad ventilation would be `3.4-=(3.4*0.8*0.1)`, which equals to an effective attack of roughly 3.1. Instead, the same outfit but on a ship with very good ventilation would be `3.4-=(3.4*0.8*0.8)`, which equals to an attack of roughly 1.2.
Secondary attributes like the previously mentioned ventable sections would probably have a similar formula, but with a a constant in the code. Something like `attack-=(0.3*attack*secondary attribute*outfit respective attribute)`, applied after the effects of the primary attributes have already been calculated

**Comment by Amazinite** [comment]
I know this isn't the exact same thing, but the idea behind it (completely overhauling boarding) seems similar to #1408. Whatever the case though, I think that MZ's comment there applies to here as well.

> This is a classic example of "feature creep." If you want to play XCom, play XCom. If you want to play FTL, play FTL. If you want to play a 4x game, play a 4x game. But all those are well beyond the scope of a space trading and combat game, which is all ES is intended to be.

That is to say, I believe that boarding combat is and should remain intentionally simplistic as that is not the main focus of the game.

**Comment by Zitchas** [comment]
I like it. 
1) This adds a small amount of complexity to boarding. It isn't much, but it is enough that there isn't one automatic "this is the best outfit to have for boarding, period."
2) I like the fact that it shouldn't have an appreciable impact on increasing the duration of boarding operations, while rewarding players for planning ahead.
3) If the outfit scanner and/or tactical scanner can be used to see what sort of modifiers a target ship has, this makes them more useful and more likely to be used, which is a good thing.
4) This gives players more choices, more diversity, and more challenge.
5) It means there are appreciable tactical differences between boarding different sorts of ships.
6) It introduces the possibility that ships designed for infantry operations (drop ships, auxillary, etc) should be better for capturing other ships than something like the StarQueen, which is expressely NOT a combat transport)

**Comment by MasterOfGrey** [comment]
@ravenshining how well it does that will depend on the numbers that are actually put in, but this setup would have the capacity to facilitate both these things.

@Amazinite I can confidently say that this is not similar to that in any sense other than that it involves the boarding action.
This preserves the current experience as it is when boarding, but simply gives a diversity of pre-planning options.

@beccabunny I do generally like that but I query it on the following points:
- What do you do if you want something to be most effective in the middle of a range when other things work from the end?
- Having two separate calculation options depending on attribute type feels cumbersome to me. Would this not then require coding them in, rather than flexibly defining attributes?

**Comment by beccabunny** [comment]
How would an implementation for flexibly defining attributes work?

**Comment by arkhne** [comment]
I think this is too ambitious in scope. While I agree that changes should be made, I think smaller steps should be made, more in the vein of reducing the effectiveness of OP options, rather than a complete overhaul that needlessly increases complexity.

**Comment by MasterOfGrey** [comment]
@beccabunny a h2h outfit could simply have a modifiers section which would contain a block of relevant modifiers:
`Narrow -0.1`
`Twisted -0.2`
`Defensive Cover -0.15`

And then if the ship you’re boarding has any of those attributes you just apply the associated modifier to the weapon effectiveness.

We can already add just any old random attribute to ships anyway, so the implementation doesn’t seem far off.

@arkhne I wouldn’t expect that it would be implemented in one go. Much of this could be broken into stages, and all the complexity remains at the outfitting stage. Once you’re in the air it’s not really any different to now.

Ultimately though, the existing system is fundamentally basic. Nerfing the current OP option only makes another option the best instead. Any change that isn’t made with a mind to overhaul how the balance works is really just putting fresh paint on a turd. Small steps are good, but pointless without an end-state vision.

**Comment by ravenshining** [comment]
Removing nerve gas would *also* remove the "why don't we just vent the ship to vacuum?" plot hole.  And I don't like the prospect of having to spend even more time in the outfitter than I already do, although it's a sacrifice I'm willing to make if it will slow down my early game capping rampage.

**Comment by Zitchas** [comment]
Also, regarding nerve gas, there's the big question of why every government that cares about the legality of things has the exact same list of what is and is not illegal.

**Comment by MasterOfGrey** [comment]
I can’t quite picture how making boarding take time would work without it being rather suicidal.

**Comment by MasterOfGrey** [comment]
Actually having looked at [#5283](https://github.com/endless-sky/endless-sky/pull/5283) I don't necessarily see that it actually helps with the fundamental balance of boarding, but I also don't see that it would incompatible with this in any way.

**Comment by VonZorch** [comment]
I have a plugin nearly completed that does add weapons and body armor. I will upload when it's complete.
But this brings back a request I made a long time ago to put subcategories in HTH outfits.  Those being "weapon", "armor", and "training". This would much simplify my current mess of defining an outfit for each combination, then requiring over a hundred missions to change outfits when you get better weapons or armor.

**Comment by MasterOfGrey** [comment]
> I have a plugin nearly completed that does add weapons and body armor. I will upload when it's complete. But this brings back a request I made a long time ago to put subcategories in HTH outfits. Those being "weapon", "armor", and "training". This would much simplify my current mess of defining an outfit for each combination, then requiring over a hundred missions to change outfits when you get better weapons or armor.

Hi VonZorch, there is a project I currently have underway to reform the boarding system according to similar lines as in this issue. 
One of the steps does involve separating those categories out. 

If you happen to have any C++ skills and want to come on the discord I'd appreciate any help I can get.

**Comment by VonZorch** [comment]
C++, yeah I can usually identify it when I see code, actually write code
not so much.
I went to Discord today, couldn't make any sense of the content.
